# 4.8 Group Data & Aggregating Varibales

# This script contains:

## 01. Import Libraies
## 02. Import Dataframes
## 03. Find aggregated mean of the order_number column group by department_id
## 04. Comment
## 05. Create a loyalty flag for existing customers
## 06. Determime if there's a difference between the spending habits of the three types of customers identified.
## 07. Create a spending flag for each user based on averaging price across all orders
## 08. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_last_order” column
## 09. Export Data

##  01. Importing Libraries

In [1]:
#Import Libraries

import pandas as pd
import numpy as np
import os

## 02. Import Dataframes

In [2]:
# Defining path
path = r'/Users/sanyamohsini/Desktop/CareerFoundry_Data Analytics/Achievement 4/04-2023 Instacart Basket Analysis'

In [3]:
path

'/Users/sanyamohsini/Desktop/CareerFoundry_Data Analytics/Achievement 4/04-2023 Instacart Basket Analysis'

In [5]:
# Importing ords_prods_merged_updated.pkl
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [6]:
# Inspect file
df_ords_prods_merged.shape

(32404859, 18)

In [8]:
df_ords_prods_merged.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   
5                   19.0         196                  1          1   
6                   20.0         196                  1          1   
7                   14.0         196                  2          1   
8                    0.0         196                  4          1   
9                   30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders  
5  Regularly busy  Regularly busy        Average orders  
6  Regularly busy    Busiest days           Most orders  
7  Regularly busy    Busiest days           Most orders  
8  Regularly busy    Busiest days           Most orders  
9      Least busy    Slowest days        Average orders

## 03. Find aggregated mean of the order_number column group by department_id

In [9]:
# Group by product_name
df_ords_prods_merged.groupby('product_name')

In [10]:
# Split data into groups by department_id and then calculate the mean values for the order_number column
df_ords_prods_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### 04. The subset has a lot more variability in the mean values. For instance, the lowest mean was 11.29 department_id 4 and the highest was 19.46 department_id 16. With the full dataset the lowest mean is 15.21 department_id 5 and the highest is 22.9 department_id 21.

## 05.  Create a loyalty flag for existing customers

In [39]:
# Group data by user_id column, apply transform function on the order_number column to generate the maximum orders for each user and create a new column max_order to place results of aggregation
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
df_ords_prods_merged.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   
20                    8.0         196                 14          1   
21                    4

In [13]:
# Create flag to assign loyalty labels based on max order value
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] <= 40) & (df_ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
# Check frequencies of loyalty flags for accuracy
df_ords_prods_merged['loyalty_flag'].value_counts(dropna= False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

## 06. Determime if there's a difference between the spending habits of the three types of customers identified.

In [19]:
# Group by prices
df_ords_prods_merged.groupby('prices')

In [22]:
#Split the data into groups by loyalty_flag and calculate mean each prices column
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

## 07. Create a spending flag for each user based on averaging price across all orders

In [38]:
# Group data by user_id column, apply transform function on the prices column to generate mean_price aggregation results
df_ords_prods_merged['mean_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [24]:
df_ords_prods_merged.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   
20                    8.0         196                 14          1   
21                    4

In [25]:
# Create flag to assign spending labels to user_id based on mean product price
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price'] < 10, 'spending_flag'] = 'Low Spender'

In [26]:
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price'] >= 10, 'spending_flag'] = 'High Spender'

In [27]:
# Checking frequencies of spending flags for accurary 
df_ords_prods_merged['spending_flag'].value_counts(dropna = False)

Low Spender     31770614
High Spender      634245
Name: spending_flag, dtype: int64

In [29]:
df_ords_prods_merged.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   
20                    8.0         196                 14          1   
21                    4

## 08. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_last_order” column.

In [40]:
# Group data by user_id column, apply transform function on the days_since_last_order column to generate median_days_last_order aggregation results
df_ords_prods_merged['median_days_last_order'] = df_ords_prods_merged.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [41]:
df_ords_prods_merged.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   
20                    8.0         196                 14          1   
21                    4

In [43]:
# Create flag to assign frequency labels to user_id based on median days since last order
df_ords_prods_merged.loc[df_ords_prods_merged['median_days_last_order'] > 20, 'frequency_flag'] = 'Non-Frequent Customer'

In [44]:
df_ords_prods_merged.loc[df_ords_prods_merged['median_days_last_order'] > 10 & (df_ords_prods_merged['median_days_last_order'] <= 20), 'frequency_flag'] = 'Regular Customer'

In [45]:
df_ords_prods_merged.loc[df_ords_prods_merged['median_days_last_order'] <= 10 , 'frequency_flag'] = 'Frequent Customer'

In [50]:
df_ords_prods_merged.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   
20                    8.0         196                 14          1   
21                    4

In [48]:
# Check frequencies of frequency flags for accuracy
df_ords_prods_merged['frequency_flag'].value_counts(dropna=False)

Frequent Customer    21559853
Regular Customer     10845001
NaN                         5
Name: frequency_flag, dtype: int64

In [51]:
# Examining the null values
print(df_ords_prods_merged[df_ords_prods_merged.frequency_flag.isnull()])

          order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_last_order  product_id  \
13645692                 17                    NaN       10749   
17251990                 17                    NaN       33401   
17622767                 17                    NaN       23695   
24138593                 17                    NaN       21334   
25880002                 17                    NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25

#### The 5 null values are all new customers, so they only have one order hence why the median days since last prior order cannot be calculated. 

## 09. Export dataframe

In [52]:
#Export dataframe as a pickle file
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_v2.pkl'))